File: SourceCode/Steel/ftt_s_scrap.py
# -*- coding: utf-8 -*-
"""
Created on Wed Feb 21 11:09:26 2024

@author: AE

=========================================
ftt_s_scrap.py
=========================================
####################################

    Calculate scrap availability in FTT-Steel.

    This function calculates the avaliability of scrap
    This is based on:

    Parameters
    -----------
    data: dictionary
        Data is a container that holds all cross-sectional (of time) for all
        variables. Variable names are keys and the values are 3D NumPy arrays.
    data_dt: dictionary
        Data_dt is a container that holds all cross-sectional (of time) data
        for all variables of the previous iteration.
    time_lag: dictionary
        Time_lag is a container that holds all cross-sectional (of time) data
        for all variables of the previous year.
    titles: dictionary
        Titles is a container of all permissible dimension titles of the model.
    dt: integer
        Dt is an integer - 1 / number of iterations.

    Returns
    ----------
    data: dictionary
        Data is a container that holds all cross-sectional (of time) data for
        all variables.
        Variable names are keys and the values are 3D NumPy arrays.
        The values inside the container are updated and returned to the main
        routine.
"""
# _______________________________________________________________________
# ============FTT:Steel model ========== Created by JF Mercure, F Knobloch, L. van Duuren, and P. Vercoulen
# Follows general equations of the FTT model (e.g. Mercure Energy Policy 2012)
#  Adapted to the iron and steel industry.
#  
# FTT determines the secnology mix
# ________________________________________jm801@cam.ac.uk________________

# ---------------Scrap availability calculation function-----------------

def scrap_calc(data, time_lag, titles, year):
     data['SSSR'] = data['SXSS'].copy()
     for r in range(len(titles['RTI'])): 
        for sec in range(len(titles['XPTI'])):
            
            Retrievedate = int(year - (data['SXLT'][r, sec, 0]))
            data['SXSC'][r, 0, 0] = 0.0
            if (year > 2017):
                data['SHS2'][r, year - 2000, 0] = data['SPSA'][r, 0, 0]
       # Calcultate Scrap availability
       # SXSCtot = SXSCpg + HSP(T- LTpg) * RRpg * SSpg * (1 - LRpg) for all pg
        if (Retrievedate < 1918):
             #If data has to be retrieved from a date at which is before the period included on the databank, then make an assumption what the steelproduction would have been
             #Assumed growth rate of 3.5%
             data['SXSC'][r, :, 0] = data['SXSC'][r, 0, 0] + data['SHS1'][r, 1, 0]*0.965**(1918 - Retrievedate) * data['SXRR'][r, sec, 0] * data['SSSR'][r, sec, 0]*(1 - (data['SXLR'][r, sec, 0]))
        elif (Retrievedate < 2017):
            data['SXSC'][r, :, 0] = data['SXSC'][r, 0, 0] + data['SHS1'][r, Retrievedate - 1917, 0] * data['SXRR'][r, sec, 0] * data['SSSR'][r, sec, 0] * (1 - data['SXLR'][r, sec, 0]) * 0.2
        if (Retrievedate+1 < 2017):
            data['SXSC'][r,: , 0] = data['SXSC'][r, 0, 0] + data['SHS1'][r, Retrievedate+0 - 1917, 0] * data['SXRR'][r, sec, 0] * data['SSSR'][r, sec, 0] * (1 - data['SXLR'][r, sec, 0]) * 0.2
        if (Retrievedate+2 < 2017):
            data['SXSC'][r,: , 0] = data['SXSC'][r, 0, 0] + data['SHS1'][r, Retrievedate+1 - 1917, 0] * data['SXRR'][r, sec, 0] * data['SSSR'][r, sec, 0] * (1 - data['SXLR'][r, sec, 0]) * 0.2
        if (Retrievedate+3 < 2017):
            data['SXSC'][r,: , 0] = data['SXSC'][r, 0, 0] + data['SHS1'][r, Retrievedate+2 - 1917, 0] * data['SXRR'][r, sec, 0] * data['SSSR'][r, sec, 0] * (1 - data['SXLR'][r, sec, 0]) * 0.15
        if (Retrievedate+4 < 2017):
            data['SXSC'][r,: , 0] = data['SXSC'][r, 0, 0] + data['SHS1'][r, Retrievedate+3 - 1917, 0] * data['SXRR'][r, sec, 0] * data['SSSR'][r, sec, 0] * (1 - data['SXLR'][r, sec, 0]) * 0.1
        if (Retrievedate+5 < 2017):
            data['SXSC'][r,: , 0] = data['SXSC'][r, 0, 0] + data['SHS1'][r, Retrievedate+4 - 1917, 0] * data['SXRR'][r, sec, 0] * data['SSSR'][r, sec, 0] * (1 - data['SXLR'][r, sec, 0]) * 0.1
        if (Retrievedate+6 < 2017):
            data['SXSC'][r,: , 0] = data['SXSC'][r, 0, 0] + data['SHS1'][r, Retrievedate+5 - 1917, 0] * data['SXRR'][r, sec, 0] * data['SSSR'][r, sec, 0] * (1 - data['SXLR'][r, sec, 0]) * 0.05          
        if (Retrievedate+1 >= 2017):
            data['SXSC'][r,: , 0] = data['SXSC'][r, 0, 0] + data['SHS2'][r, Retrievedate+0 - 2000, 0] * data['SXRR'][r, sec, 0] * data['SSSR'][r, sec, 0] * (1 - data['SXLR'][r, sec, 0]) * 0.2
        if (Retrievedate+2 >= 2017):
            data['SXSC'][r,: , 0] = data['SXSC'][r, 0, 0] + data['SHS2'][r, Retrievedate+1 - 2000, 0] * data['SXRR'][r, sec, 0] * data['SSSR'][r, sec, 0] * (1 - data['SXLR'][r, sec, 0]) * 0.2
        if (Retrievedate+3 >= 2017):
            data['SXSC'][r,: , 0] = data['SXSC'][r, 0, 0] + data['SHS2'][r, Retrievedate+2 - 2000, 0] * data['SXRR'][r, sec, 0] * data['SSSR'][r, sec, 0] * (1 - data['SXLR'][r, sec, 0]) * 0.15
        if (Retrievedate+4 >= 2017):
            data['SXSC'][r,: , 0] = data['SXSC'][r, 0, 0] + data['SHS2'][r, Retrievedate+3 - 2000, 0] * data['SXRR'][r, sec, 0] * data['SSSR'][r, sec, 0] * (1 - data['SXLR'][r, sec, 0]) * 0.1
        if (Retrievedate+5 >= 2017):
            data['SXSC'][r,: , 0] = data['SXSC'][r, 0, 0] + data['SHS2'][r, Retrievedate+4 - 2000, 0] * data['SXRR'][r, sec, 0] * data['SSSR'][r, sec, 0] * (1 - data['SXLR'][r, sec, 0]) * 0.1
        if (Retrievedate+6 >= 2017):
            data['SXSC'][r,: , 0] = data['SXSC'][r, 0, 0] + data['SHS2'][r, Retrievedate+5 - 2000, 0] * data['SXRR'][r, sec, 0] * data['SSSR'][r, sec, 0] * (1 - data['SXLR'][r, sec, 0]) * 0.05
        #If endogenous sector split calculation is going to be included then it should be fed in here to replace SXSS
        elif (Retrievedate >= 2017):
            data['SXSC'][r,: , 0] = data['SXSC'][r, 0, 0] + data['SHS2'][r, Retrievedate - 2000, 0] * data['SXRR'][r, sec, 0] * data['SSSR'][r, sec, 0] * (1 - data['SXLR'][r, sec, 0]) * 0.2
            data['SXSC'][r,: , 0] = data['SXSC'][r, 0, 0] + data['SHS2'][r, Retrievedate+0 - 2000, 0] * data['SXRR'][r, sec, 0] * data['SSSR'][r, sec, 0] * (1 - data['SXLR'][r, sec, 0]) * 0.2
            data['SXSC'][r,: , 0] = data['SXSC'][r, 0, 0] + data['SHS2'][r, Retrievedate+1 - 2000, 0] * data['SXRR'][r, sec, 0] * data['SSSR'][r, sec, 0] * (1 - data['SXLR'][r, sec, 0]) * 0.2
            data['SXSC'][r,: , 0] = data['SXSC'][r, 0, 0] + data['SHS2'][r, Retrievedate+2 - 2000, 0] * data['SXRR'][r, sec, 0] * data['SSSR'][r, sec, 0] * (1 - data['SXLR'][r, sec, 0]) * 0.15
            data['SXSC'][r,: , 0] = data['SXSC'][r, 0, 0] + data['SHS2'][r, Retrievedate+3 - 2000, 0] * data['SXRR'][r, sec, 0] * data['SSSR'][r, sec, 0] * (1 - data['SXLR'][r, sec, 0]) * 0.1
            data['SXSC'][r,: , 0] = data['SXSC'][r, 0, 0] + data['SHS2'][r, Retrievedate+4 - 2000, 0] * data['SXRR'][r, sec, 0] * data['SSSR'][r, sec, 0] * (1 - data['SXLR'][r, sec, 0]) * 0.1
            data['SXSC'][r,: , 0] = data['SXSC'][r, 0, 0] + data['SHS2'][r, Retrievedate+5 - 2000, 0] * data['SXRR'][r, sec, 0] * data['SSSR'][r, sec, 0] * (1 - data['SXLR'][r, sec, 0]) * 0.05

     return data

File: FTT-S-Scrap.f90
!_______________________________________________________________________
!_______________________________________________________________________
! ============FTT:Steel model ========== Created by JF Mercure, F Knobloch, L. van Duuren, and P. Vercoulen
! Follows general equations of the FTT model (e.g. Mercure Energy Policy 2012)
! Adapted to the iron and steel industry.
! 
! FTT determines the technology mix
!________________________________________jm801@cam.ac.uk________________

!---------------Scrap availability calculation function-----------------
    
    SUBROUTINE IDFTTSScrap(NST,NXP,NR,NC5,NY1,NSM,NY,NQ,NBQRY,SHS1,SHS2,SXSC,BSTC,SSSR, SXSS,SXLT,SXLR,SXRR,SLT2,SPSA,SXS1,SXS2,SXS3,SXS4,BQRY,IRUN,DATE)
    IMPLICIT NONE
    
    INTEGER,INTENT(IN) :: NST,NXP,NR,NC5,NY1,NSM,NY,NQ,NBQRY
    INTEGER,INTENT(IN) :: IRUN,DATE
    
    REAL , INTENT(INOUT), DIMENSION(NR, NY1) :: SHS1, SHS2   !Historical steel production for the period of 1918-1917, and future steel production (2000-2100). Used to calculated scrap availability
    REAL , INTENT(INOUT), DIMENSION(NR, NY1) :: SXS1,SXS2,SXS3,SXS4 !Endogenous calculation of sector splits.
    REAL , INTENT(INOUT), DIMENSION(NR) :: SPSA
    REAL , INTENT(OUT), DIMENSION(NR) :: SXSC
    REAL , INTENT(INOUT), DIMENSION(NST,NR,NC5) :: BSTC
    REAL , INTENT(INOUT), DIMENSION(NR,NXP) :: SSSR, SXSS, SXLT, SXLR, SXRR, SLT2
    REAL , DIMENSION(NQ,NR,NBQRY) :: BQRY
    INTEGER , DIMENSION(NY) :: non_EU_mask, EU_mask
    REAL , DIMENSION(NY,NR) :: SteelUse,SteelUse0,SteelUse1,SteelUse2,SteelUse3,SteelUse4
    REAL , DIMENSION(NR) :: TotalSteelUse
    REAL :: TestB1_1, TestB1_2, TestB2_1, TestB2_2, TEST_SPLIT, TEST_TOTAL
    INTEGER :: J, PG, RetrieveDate, I
    SSSR = SXSS
    IF (DATE > 2017) SHS2(:,DATE - 2000) = SPSA
    SXSC = 0.0
    non_EU_mask =   (/ 3,3,3,2,3,3,3,3,3,3,3,3,2,0,0,3,4,1,1,2,4,3,3,3,3,1,4,4,3,1,1,3,4,4,4,4,4,4,4,4,4,4,2,4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0 /)
    EU_mask =       (/ 3,3,1,3,3,3,3,3,3,3,3,3,3,3,3,3,0,0,4,4,2,1,1,4,3,3,3,3,3,3,3,3,3,3,1,1,3,1,3,4,3,3,4,4,4,4,3,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4 /)
    DO J = 1,NR
        DO PG = 1, NXP
            RetrieveDate = NINT(DATE - SXLT(J,PG)) 
            
            !Calculate scrap availability
            !SXSCtot = SXSCpg + HSP(T-LTpg) * RRpg * SSpg * (1-LRpg) for all pg
            IF (RetrieveDate < 1918) THEN
                !If data has to be retrieved from a date at which is before the period included on the databank, then make an assumption what the steelproduction would have been
                !Assumed growth rate of 3.5%
                SXSC(J) = SXSC(J) + SHS1(J,1)*(0.965)**(1918 - RetrieveDate) * SXRR(J,PG) * SSSR(J,PG) * (1-SXLR(J,PG))
            ELSEIF (RetrieveDate < 2017) THEN
                SXSC(J) = SXSC(J) + SHS1(J,RetrieveDate - 1917) * SXRR(J,PG) * SSSR(J,PG) * (1-SXLR(J,PG)) * 0.2
                IF (RetrieveDate+1 < 2017) SXSC(J) = SXSC(J) + SHS1(J,RetrieveDate+1 - 1917) * SXRR(J,PG) * SSSR(J,PG) * (1-SXLR(J,PG)) * 0.2
                IF (RetrieveDate+2 < 2017) SXSC(J) = SXSC(J) + SHS1(J,RetrieveDate+2 - 1917) * SXRR(J,PG) * SSSR(J,PG) * (1-SXLR(J,PG)) * 0.2
                IF (RetrieveDate+3 < 2017) SXSC(J) = SXSC(J) + SHS1(J,RetrieveDate+3 - 1917) * SXRR(J,PG) * SSSR(J,PG) * (1-SXLR(J,PG)) * 0.15
                IF (RetrieveDate+4 < 2017) SXSC(J) = SXSC(J) + SHS1(J,RetrieveDate+4 - 1917) * SXRR(J,PG) * SSSR(J,PG) * (1-SXLR(J,PG)) * 0.1
                IF (RetrieveDate+5 < 2017) SXSC(J) = SXSC(J) + SHS1(J,RetrieveDate+5 - 1917) * SXRR(J,PG) * SSSR(J,PG) * (1-SXLR(J,PG)) * 0.1
                IF (RetrieveDate+6 < 2017) SXSC(J) = SXSC(J) + SHS1(J,RetrieveDate+6 - 1917) * SXRR(J,PG) * SSSR(J,PG) * (1-SXLR(J,PG)) * 0.05
                IF (RetrieveDate+1 .GE. 2017) SXSC(J) = SXSC(J) + SHS2(J,RetrieveDate+1 - 2000) * SXRR(J,PG) * SSSR(J,PG) * (1-SXLR(J,PG)) * 0.2
                IF (RetrieveDate+2 .GE. 2017) SXSC(J) = SXSC(J) + SHS2(J,RetrieveDate+2 - 2000) * SXRR(J,PG) * SSSR(J,PG) * (1-SXLR(J,PG)) * 0.2
                IF (RetrieveDate+3 .GE. 2017) SXSC(J) = SXSC(J) + SHS2(J,RetrieveDate+3 - 2000) * SXRR(J,PG) * SSSR(J,PG) * (1-SXLR(J,PG)) * 0.15
                IF (RetrieveDate+4 .GE. 2017) SXSC(J) = SXSC(J) + SHS2(J,RetrieveDate+4 - 2000) * SXRR(J,PG) * SSSR(J,PG) * (1-SXLR(J,PG)) * 0.1
                IF (RetrieveDate+5 .GE. 2017) SXSC(J) = SXSC(J) + SHS2(J,RetrieveDate+5 - 2000) * SXRR(J,PG) * SSSR(J,PG) * (1-SXLR(J,PG)) * 0.1
                IF (RetrieveDate+6 .GE. 2017) SXSC(J) = SXSC(J) + SHS2(J,RetrieveDate+6 - 2000) * SXRR(J,PG) * SSSR(J,PG) * (1-SXLR(J,PG)) * 0.05
            !If endogenous sector split calculation is going to be included then it should be fed in here to replace SXSS
            ELSEIF (RetrieveDate .GE. 2017) THEN
                SXSC(J) = SXSC(J) + SHS2(J,RetrieveDate - 2000) * SXRR(J,PG) * SSSR(J,PG) * (1-SXLR(J,PG)) * 0.2
                SXSC(J) = SXSC(J) + SHS2(J,RetrieveDate+1 - 2000) * SXRR(J,PG) * SSSR(J,PG) * (1-SXLR(J,PG)) * 0.2
                SXSC(J) = SXSC(J) + SHS2(J,RetrieveDate+2 - 2000) * SXRR(J,PG) * SSSR(J,PG) * (1-SXLR(J,PG)) * 0.2
                SXSC(J) = SXSC(J) + SHS2(J,RetrieveDate+3 - 2000) * SXRR(J,PG) * SSSR(J,PG) * (1-SXLR(J,PG)) * 0.15
                SXSC(J) = SXSC(J) + SHS2(J,RetrieveDate+4 - 2000) * SXRR(J,PG) * SSSR(J,PG) * (1-SXLR(J,PG)) * 0.1
                SXSC(J) = SXSC(J) + SHS2(J,RetrieveDate+5 - 2000) * SXRR(J,PG) * SSSR(J,PG) * (1-SXLR(J,PG)) * 0.1
                SXSC(J) = SXSC(J) + SHS2(J,RetrieveDate+6 - 2000) * SXRR(J,PG) * SSSR(J,PG) * (1-SXLR(J,PG)) * 0.05
                
            ENDIF
        ENDDO
    ENDDO
END
    
            
            
        